File: video-series-command/src/main/java/org/jesperancinha/video/command/aggregates/VideoSeriesAggregate.java
package org.jesperancinha.video.command.aggregates;

import javax.persistence.Entity;
import javax.persistence.Id;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.axonframework.commandhandling.CommandHandler;
import org.axonframework.eventsourcing.EventSourcingHandler;
import org.axonframework.modelling.command.AggregateIdentifier;
import org.axonframework.spring.stereotype.Aggregate;
import org.jesperancinha.video.command.commands.AddVideoSeriesCommand;
import org.jesperancinha.video.core.events.AddSeriesEvent;
import org.springframework.stereotype.Component;

import java.util.UUID;

import static org.axonframework.modelling.command.AggregateLifecycle.apply;

@Data
@Aggregate
@Entity
@NoArgsConstructor
public class VideoSeriesAggregate {

    @AggregateIdentifier
    @Id
    private String id;

    @CommandHandler
    public VideoSeriesAggregate(AddVideoSeriesCommand command) {
        apply(AddSeriesEvent.builder()
                .id(UUID.randomUUID().toString())
                .cashValue(command.getCashValue())
                .genre(command.getGenre())
                .name(command.getName())
                .volumes(command.getVolumes()).build()
        );
    }

    @EventSourcingHandler
    public void on(AddSeriesEvent event) {
        this.id = event.getId();
    }

}

File: video-series-command/src/main/kotlin/org/jesperancinha/video/command/aggregates/VideoSeriesAggregate.kt
package org.jesperancinha.video.command.aggregates

import org.axonframework.commandhandling.CommandHandler
import org.axonframework.eventsourcing.EventSourcingHandler
import org.axonframework.modelling.command.AggregateIdentifier
import org.axonframework.modelling.command.AggregateLifecycle
import org.axonframework.spring.stereotype.Aggregate
import org.jesperancinha.video.command.commands.AddVideoSeriesCommand
import org.jesperancinha.video.core.data.Genre
import org.jesperancinha.video.core.events.AddSeriesEvent
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.math.BigDecimal
import java.util.*

@Aggregate
class VideoSeriesAggregate {
    @AggregateIdentifier
    var id: String? = null
    var name: String? = null
    var volumes: Int? = null
    var cashValue: BigDecimal? = null
    var genre: Genre? = null

    constructor()

    @CommandHandler
    constructor(command: AddVideoSeriesCommand) {
        logger.info("Created addVideoSeriesCommand {}", command)
        val event =             AddSeriesEvent(
            id = UUID.randomUUID().toString(),
            cashValue = requireNotNull(command.cashValue),
            genre = requireNotNull(command.genre),
            name = requireNotNull(command.name),
            volumes = requireNotNull(command.volumes)
        )
        AggregateLifecycle.apply(event)
    }

    @EventSourcingHandler
    fun on(event: AddSeriesEvent) {
        id = event.id
        name = event.name
        volumes = event.volumes
        cashValue = event.cashValue
        genre = event.genre
    }

    companion object {
        val logger: Logger = LoggerFactory.getLogger(VideoSeriesAggregate::class.java)
    }
}